codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedIntegerGenerator.kt

```diff
@@ -0,0 +1,187 @@
+ * the constraints.
+ */
+class ConstrainedIntegerGenerator(
+    val codegenContext: ServerCodegenContext,
+    val writer: RustWriter,
+    val shape: IntegerShape,
+) {
+    val model = codegenContext.model
+    val constrainedShapeSymbolProvider = codegenContext.constrainedShapeSymbolProvider
+    val publicConstrainedTypes = codegenContext.settings.codegenConfig.publicConstrainedTypes
+    private val constraintViolationSymbolProvider =
+        with(codegenContext.constraintViolationSymbolProvider) {
+            if (publicConstrainedTypes) {
+                this
+            } else {
+                PubCrateConstraintViolationSymbolProvider(this)
+            }
+        }
+
+    fun render() {
+        val rangeTrait = shape.expectTrait<RangeTrait>()
+
+        val symbol = constrainedShapeSymbolProvider.toSymbol(shape)
+        val name = symbol.name
+        val inner = RustType.Integer(32).render()
+        val constraintViolation = constraintViolationSymbolProvider.toSymbol(shape)
+
+        val condition = if (rangeTrait.min.isPresent && rangeTrait.max.isPresent) {
+            "(${rangeTrait.min.get()}..=${rangeTrait.max.get()}).contains(&value)"
+        } else if (rangeTrait.min.isPresent) {
+            "${rangeTrait.min.get()} <= value"
+        } else {
+            "value <= ${rangeTrait.max.get()}"
+        }
+
+        val constrainedTypeVisibility = if (publicConstrainedTypes) {
+            Visibility.PUBLIC
+        } else {
+            Visibility.PUBCRATE
+        }
+        val constrainedTypeMetadata = RustMetadata(
+            Attribute.Derives(setOf(RuntimeType.Debug, RuntimeType.Clone, RuntimeType.PartialEq, RuntimeType.Eq, RuntimeType.Hash)),
+            visibility = constrainedTypeVisibility,
+        )
+
+        writer.documentShape(shape, model, note = rustDocsNote(name))
+        constrainedTypeMetadata.render(writer)
+        writer.rust("struct $name(pub(crate) $inner);")
+        if (constrainedTypeVisibility == Visibility.PUBCRATE) {
+            Attribute.AllowUnused.render(writer)
+        }
+        writer.rustTemplate(
+            """
+            impl $name {
+                /// ${rustDocsInnerMethod(inner)}
+                pub fn inner(&self) -> &$inner {
+                    &self.0
+                }
+
+                /// ${rustDocsIntoInnerMethod(inner)}
+                pub fn into_inner(self) -> $inner {
+                    self.0
+                }
+            }
+
+            impl #{ConstrainedTrait} for $name  {
+                type Unconstrained = $inner;
+            }
+
+            impl #{From}<$inner> for #{MaybeConstrained} {
+                fn from(value: $inner) -> Self {
+                    Self::Unconstrained(value)
+                }
+            }
+
+            impl #{From}<$name> for #{MaybeConstrained} {
+                fn from(value: $name) -> Self {
+                    Self::Constrained(value)
+                }
+            }
```

Comment:
We don't need this one do we?

Reply:
No, we don't - removed in https://github.com/awslabs/smithy-rs/pull/2005/commits/d71932886636f7c7a8fecdd62d836d5143f32c55